### PR TITLE
FormItemView refactoring

### PR DESCRIPTION
**Pull request by descorp**
### Changes

* refactor FormValueItem as open class
* refactor Value of FormValueItem to be observable and deprecate delegates
* make base class for  FormPickerItem
* make FormPhoneExtensionPickerItem inherit from BaseFormPickerItem
* make FormComponentStyle a FormValueItemStyle and refactor `tintColor` property
* simplify FormViewItemManager

### CODE REVIEW

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+/// Describes a picker item.
+open class BaseFormValuePickerItem<T: CustomStringConvertible>: FormValueItem<BasePickerElement<T>, FormTextItemStyle>,
+    InputViewRequiringFormItem {
+
+    /// The complete list of selectable values.
+    internal let selectableValues: [BasePickerElement<T>]
+
+    /// The title of the item.
+    public var title: String?
```
Comment (kaphacius) [inline Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:45]: can we make it a constant and pass as a parameter to the initializer?

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    internal func updateSelection() {
+        inputControl.label = item.value.wrappedValue.description
+    }
+
+    internal func initialize() {
+        addSubview(inputControl)
+        inputControl.translatesAutoresizingMaskIntoConstraints = false
+        inputControl.preservesSuperviewLayoutMargins = true
+        (inputControl as UIView).adyen.anchor(inside: self)
```
Comment (kaphacius) [inline Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift:61]: can we do it without casting here?

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
+        guard row < item.selectableValues.count else { return }
+        didSelection(selectionValue: item.selectableValues[row])
+    }
+
+    // MARK: - Private
+
+    private func didSelection(selectionValue: BasePickerElement<T>) {
```
Comment (kaphacius) [inline Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift:104]: maybe not the best name. `didSelect` or `onSelect`?

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BasePickerInputControl.swift [hunk 4]
```diff
@@ -1,56 +1,87 @@
...
+    var onDidResignFirstResponder: (() -> Void)? { get set }
+
+    /// Executed when the view becomes first responder.
+    var onDidBecomeFirstResponder: (() -> Void)? { get set }
+
+    /// Executed when the view detected tap.
+    var onDidTap: (() -> Void)? { get set }
+
+    /// Controls visibility of chevrone view.
+    var showChevrone: Bool { get set }
```
Comment (kaphacius) [inline Adyen/UI/Form/Items/Value Pickers/Abstract/BasePickerInputControl.swift:21]: *showChevron :)

File: Adyen/UI/UI Style/FormComponentStyle.swift [hunk 5]
```diff
@@ -8,10 +8,15 @@ import Foundation
...
 
 /// Contains the styling customization options for any form-based component.
-public struct FormComponentStyle: ViewStyle {
+public struct FormComponentStyle: FormValueItemStyle {
     
     /// :nodoc:
     public var backgroundColor = UIColor.Adyen.componentBackground
+
+    /// The section header style.
+    public var sectionHeader = TextStyle(font: .preferredFont(forTextStyle: .headline),
```
Comment (kaphacius) [inline Adyen/UI/UI Style/FormComponentStyle.swift:17]: can this be a constant? maybe have one designated initializer where fields are set to passed (or default) values. and a few convenience initializers, that call designated with default parameters

File: Adyen/UI/Form/FormItemViewBuilder.swift [hunk 6]
```diff
@@ -30,7 +30,7 @@ public struct FormItemViewBuilder {
     
     /// Builds `FormPhoneExtensionPickerItemView` from `FormPhoneExtensionPickerItem`.
     /// :nodoc:
-    internal func build(with item: FormPhoneExtensionPickerItem) -> FormItemView<FormPhoneExtensionPickerItem> {
+    public func build(with item: FormPhoneExtensionPickerItem) -> BaseFormValuePickerItemView<PhoneExtensionViewModel> {
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/FormItemViewBuilder.swift:33]: `BaseFormPickerItemView` name is better, the word `value` is redundant and makes the name long.

File: Adyen/UI/Form/Items/Text/FormTextItem.swift [hunk 7]
```diff
@@ -9,207 +9,43 @@ import UIKit
...
-        }
-        set {
-            objc_setAssociatedObject(self,
-                                     &FormTextItemAssociatedKeys.validationFailureMessage,
-                                     newValue,
-                                     objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
-        }
+
+    override open func build(with builder: FormItemViewBuilder) -> AnyFormItemView {
+        preconditionFailure("This is abstract method. Override it on high tier object")
```
[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/UI/Form/Items/Text/FormTextItem.swift:48]: `fatalError` makes more sense.

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    public static func == (lhs: Self, rhs: Self) -> Bool {
+        lhs.identifier == rhs.identifier
+    }
+
+    public var description: String { item.description }
+
+}
+
+/// Describes a picker item.
+open class BaseFormValuePickerItem<T: CustomStringConvertible>: FormValueItem<BasePickerElement<T>, FormTextItemStyle>,
```
Comment (mohammedDehairy) [inline Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:38]: `BaseFormPickerItem` is enough, the word `Value` is not needed per-se I think.

File: Adyen/UI/UI Style/FormComponentStyle.swift [hunk 9]
```diff
@@ -8,10 +8,15 @@ import Foundation
 import UIKit
 
 /// Contains the styling customization options for any form-based component.
-public struct FormComponentStyle: ViewStyle {
+public struct FormComponentStyle: FormValueItemStyle {
```
Comment (mohammedDehairy) [inline Adyen/UI/UI Style/FormComponentStyle.swift:11]: why does `FormComponentStyle` inherit from `FormValueItemStyle`? the former is for the while form, the second is for individual value fields, this makes little sense to me.

File: Adyen/UI/Form/FormViewItemManager.swift [hunk 10]
```diff
@@ -70,9 +68,6 @@ internal final class FormViewItemManager {
     
     private func newItemView<T: FormItem>(for item: T) -> AnyFormItemView {
         let itemView = item.build(with: FormItemViewBuilder())
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/FormViewItemManager.swift:70]: Was debugging something.. 
👍

File: Adyen/UI/Form/FormViewItemManager.swift [hunk 11]
```diff
@@ -30,13 +25,16 @@ internal final class FormViewItemManager {
     ///   - item: The item to append.
     ///   - itemViewType: Optionally, the item view type to use for this item.
     ///                   When none is specified, the default will be used.
-    internal func append<T: FormItem>(_ item: T) {
+    @discardableResult
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/FormViewItemManager.swift:28]: 👍

File: Adyen/UI/Form/Items/Text/FormTextItem.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/Items/Text/FormTextItem.swift:48]: 👍

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    /// :nodoc:
+    var identifier: String { get }
+}
+
+/// :nodoc:
+public struct BasePickerElement<T: CustomStringConvertible>: PickerElement {
+
+    /// Picker item identifier.
+    /// :nodoc:
+    public var identifier: String
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:23]: why not 👍

[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:23]: 👍

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// Picker item value.
+    /// :nodoc:
+    public var item: T
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:27]: 👍

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+    public static func == (lhs: Self, rhs: Self) -> Bool {
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:29]: Good idea!

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift (hunk 1 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:45]: This one is actually not used 🙊

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift (hunk 2 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift:61]: No unfortunately - inputControl is generic and Swift complains:
```Member 'adyen' cannot be used on value of protocol type 'PickerTextInputControl'; use a generic constraint instead```

This is a soft casting, tho

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift (hunk 3 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormValuePickerItemView.swift:104]: Used to be a delegate here. 👍

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BasePickerInputControl.swift (hunk 4 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BasePickerInputControl.swift:21]: 👍

File: Adyen/UI/UI Style/FormComponentStyle.swift (hunk 5 above)
Comment (descorp) [reply Adyen/UI/UI Style/FormComponentStyle.swift:17]: It should be, sure.
But we better do it in other PR 😝

File: Adyen/UI/Form/FormItemViewBuilder.swift (hunk 6 above)
Comment (descorp) [reply Adyen/UI/Form/FormItemViewBuilder.swift:33]: 👍

File: Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift (hunk 8 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/Abstract/BaseFormPickerItem.swift:38]: 👍  👍

File: Adyen/UI/UI Style/FormComponentStyle.swift (hunk 9 above)
Comment (descorp) [reply Adyen/UI/UI Style/FormComponentStyle.swift:11]: I see your point.
I will make it just tintable

**Review by kaphacius — APPROVED** [review]

**Review by kaphacius — APPROVED** [review]

**Review by kaphacius — APPROVED** [review]

**Pull request merged**